torch/csrc/jit/topological_index.h

```diff
@@ -0,0 +1,168 @@
+      const auto newIndex = nodeToIndex_[insertPoint] - defaultInterval_;
+      nodeToIndex_[toInsert] = newIndex;
+      indexToNode_[newIndex] = toInsert;
+    } else {
+      // We are between two nodes. Find the previous one
+      indexIter--;
+      const int64_t prevIndex = indexIter->first;
+      const int64_t indexBetween = (insertIndex + prevIndex) / 2;
+      if (indexToNode_.count(indexBetween) != 0) {
+        reIndex();
+        return insertBefore(insertPoint, toInsert);
+      }
+
+      nodeToIndex_[toInsert] = indexBetween;
+      indexToNode_[indexBetween] = toInsert;
+    }
+  }
+
+  // Insert `toInsert` before `insertPoint` in the topological index
+  void insertAfter(const Node* insertPoint, const Node* toInsert) {
+    // Can't insert a node twice
+    AT_ASSERT(nodeToIndex_.count(toInsert) == 0);
+
+    if (insertPoint == output_) {
+      // inserting AFTER the head/output node means PREPEND to the graph
+      return insertAfter(input_, toInsert);
+    }
+
+    auto indexIter = indexToNode_.find(nodeToIndex_[insertPoint]);
+    const auto insertIndex = indexIter->first;
+    // Are we at the end?
+    if (indexIter->second != indexToNode_.rbegin()->second) {
+      // if not, just insert before the next node
+      insertBefore((++indexIter)->second, toInsert);
+    } else {
+      // check we're not running off the end
+      if (insertIndex > (upperBound_ - defaultInterval_)) {
+        reIndex();
+        return insertAfter(insertPoint, toInsert);
+      }
+
+      // Move up a suitably vast distance and add the node
+      const auto newIndex = nodeToIndex_[insertPoint] + defaultInterval_;
+      nodeToIndex_[toInsert] = newIndex;
+      indexToNode_[newIndex] = toInsert;
+    }
+  }
+
+  // Erase `toErase` from the topological index
+  void erase(const Node* toErase) {
+    indexToNode_.erase(nodeToIndex_.at(toErase));
+    nodeToIndex_.erase(toErase);
+  }
+
+ private:
+  // If we run out of space between nodes we need to rebuild the index and
+  // "spread out" the nodes again.
+  void reIndex() {
+    AT_ASSERT(defaultInterval_ * (int64_t)nodeToIndex_.size() < upperBound_);
+
+    std::map<int64_t, const Node*> newIndexToNode;
+    std::unordered_map<const Node*, int64_t> newNodeToIndex;
+    int64_t curPos = 0;
+    for (const auto pr : indexToNode_) {
+      newIndexToNode[curPos] = pr.second;
+      newNodeToIndex[pr.second] = curPos;
+      curPos += defaultInterval_;
+    }
+    indexToNode_.swap(newIndexToNode);
+    nodeToIndex_.swap(newNodeToIndex);
+  }
+
+  // The node list is implemented as a circular linked list, with the output
+  // node as the head/tail. Therefore, the index needs to treat "appends" to the
+  // output as appends to the input.
+  const Node* input_;
+  const Node* output_;
+
+  const int64_t defaultInterval_;
+  const int64_t lowerBound_;
+  const int64_t upperBound_;
+
+  std::unordered_map<const Node*, int64_t> nodeToIndex_;
+  std::map<int64_t, const Node*> indexToNode_;
```

Comment:
I think we should just add `int64_t topoindex` to Node*.  This would prevent needing both of the maps (indexToNode is only used to get an iterator), and would allow us to delete eraseNode and a ton of logic keeping these maps consistent. It is not like this data structure isn't already internal in the graph as both Node* and Block* have modifications using it.

Reply:
Sure, maybe I'll do something similar to graph_node_list